File: Python/2017/13.py
from lib import *

input = read_input(2017, 13)

lines = input.splitlines()


R = {}
S = {}
for line in lines:
    a, b = map(int, line.split(": "))
    R[a] = b
    S[a] = 0

n = max(S)
out = 0
for i in range(n):
    if i in S and S[i] == 0:
        out += i * R[i]

    for k, v in S.items():
        S[k] = (v + 1) % (R[k] * 2 - 2)

print(out)


def step(R, S):
    for k, v in S.items():
        S[k] = (v + 1) % (R[k] * 2 - 2)


R = {}
SC = {}
for line in lines:
    a, b = map(int, line.split(": "))
    R[a] = b
    SC[a] = 0

n = max(SC)


def test():
    S = SC.copy()
    for i in range(n + 1):
        if i in S and S[i] == 0:
            return False
        step(R, S)
    return True


k = 0
while not test():
    k += 1
    step(R, SC)

print(k)

File: Rust/2023/13.rs
#![feature(test)]

type Input = Vec<Grid>;

type Grid = Vec<Vec<bool>>;

fn setup(input: &str) -> Input {
    input
        .split("\n\n")
        .map(|block| {
            block
                .lines()
                .map(|line| line.bytes().map(|b| b == b'#').collect())
                .collect()
        })
        .collect()
}

#[derive(Debug, Clone, Copy)]
enum Reflection {
    Vertical(usize),
    Horizontal(usize),
}

impl Reflection {
    fn generate_for(grid: &Grid) -> impl Iterator<Item = Self> {
        (1..grid.len())
            .map(Self::Horizontal)
            .chain((1..grid[0].len()).map(Self::Vertical))
    }

    fn count_mismatches(self, grid: &Grid) -> usize {
        match self {
            Reflection::Vertical(j) => (0..grid.len())
                .map(|i| {
                    let len = j.min(grid[0].len() - j);
                    (j - len..j)
                        .rev()
                        .zip(j..j + len)
                        .filter(|&(j1, j2)| grid[i][j1] != grid[i][j2])
                        .count()
                })
                .sum(),
            Reflection::Horizontal(i) => (0..grid[0].len())
                .map(|j| {
                    let len = i.min(grid.len() - i);
                    (i - len..i)
                        .rev()
                        .zip(i..i + len)
                        .filter(|&(i1, i2)| grid[i1][j] != grid[i2][j])
                        .count()
                })
                .sum(),
        }
    }

    fn summarize(self) -> usize {
        match self {
            Self::Vertical(j) => j,
            Self::Horizontal(i) => 100 * i,
        }
    }
}

fn solve<const N: usize>(input: &Input) -> usize {
    input
        .iter()
        .map(|grid| {
            Reflection::generate_for(grid)
                .find(|reflection| dbg!(reflection.count_mismatches(grid)) == N)
                .unwrap()
                .summarize()
        })
        .sum()
}

fn part1(input: &Input) -> usize {
    solve::<0>(input)
}

fn part2(input: &Input) -> usize {
    solve::<1>(input)
}

aoc::main!(2023, 13, ex: 1);
